File: packages/proto/ipc.ts
/** Protocol version, useful for migrations */
export const HOP_IPC_VERSION = 1 as const;

/** High-level message envelope */
export type HopMessage =
  | HopRequestMessage
  | HopResponseMessage
  | HopNotificationMessage;

/** Every message has a direction + id (except fire-and-forget notifications) */
export interface HopBaseMessage {
  /** Protocol version */
  v: typeof HOP_IPC_VERSION;
}

/** Client → Server request */
export interface HopRequestMessage extends HopBaseMessage {
  kind: 'request';
  /** Correlation id; echoed back in the response */
  id: string;
  /** The actual operation */
  request: HopRequest;
}

/** Server → Client response */
export interface HopResponseMessage extends HopBaseMessage {
  kind: 'response';
  /** Must match the request id */
  id: string;
  /** Result of executing a HopRequest */
  response: HopResponse;
}

/** Server → Client (or client → server) push events */
export interface HopNotificationMessage extends HopBaseMessage {
  kind: 'notification';
  /** Unidirectional event, no response expected */
  event: HopEvent;
}

/* ------------------------------------------------------------------ */
/* Requests                                                           */
/* ------------------------------------------------------------------ */

export type HopRequest =
  | HopFsReadRequest
  | HopFsWriteRequest
  | HopFsDeleteRequest
  | HopFsSearchRequest
  | HopWorkspaceOpenRequest
  | HopWorkspaceListRequest
  | HopTerminalSpawnRequest
  | HopTerminalWriteRequest
  | HopTerminalResizeRequest
  | HopTerminalKillRequest
  | HopLspRequest;

export interface HopFsReadRequest {
  type: 'fs.read';
  path: string;
  root?: string;
}

export interface HopFsWriteRequest {
  type: 'fs.write';
  path: string;
  content: string;
  root?: string;
}

export interface HopFsDeleteRequest {
  type: 'fs.delete';
  path: string;
  root?: string;
}

export interface HopFsSearchRequest {
  type: 'fs.search';
  query: string;
  root?: string;
}

export interface HopWorkspaceOpenRequest {
  type: 'workspace.open';
  root: string;
}

export interface HopWorkspaceListRequest {
  type: 'workspace.list';
  root: string;
}

export interface HopTerminalSpawnRequest {
  type: 'terminal.spawn';
  /** User-friendly terminal id (client-side) */
  id: string;
  /** Optional shell binary (e.g. /bin/bash, powershell.exe) */
  shell?: string;
}

export interface HopTerminalWriteRequest {
  type: 'terminal.write';
  id: string;
  data: string;
}

export interface HopTerminalResizeRequest {
  type: 'terminal.resize';
  id: string;
  cols: number;
  rows: number;
}

export interface HopTerminalKillRequest {
  type: 'terminal.kill';
  id: string;
  signal?: string;
}

export interface HopLspRequest {
  type: 'lsp.request';
  /** LSP server identifier, e.g. "tsserver", "pyright" */
  server: string;
  /** Raw LSP payload (JSON-RPC) */
  payload: unknown;
}

/* ------------------------------------------------------------------ */
/* Responses (typed per operation)                                    */
/* ------------------------------------------------------------------ */

/**
 * Top-level response union per request type.
 * You always wrap these inside HopResponseMessage.
 */
export type HopResponse =
  | HopFsReadResponse
  | HopFsWriteResponse
  | HopFsDeleteResponse
  | HopFsSearchResponse
  | HopWorkspaceOpenResponse
  | HopWorkspaceListResponse
  | HopTerminalSpawnResponse
  | HopTerminalWriteResponse
  | HopTerminalResizeResponse
  | HopTerminalKillResponse
  | HopLspResponse
  | HopGenericErrorResponse;

/** Base success/failure discriminant */
interface HopBaseResponse {
  /** The original request type this responds to (for debugging) */
  type: HopRequest['type'] | 'error';
}

export interface HopFsReadResponse extends HopBaseResponse {
  type: 'fs.read';
  ok: boolean;
  /** File content on success */
  content?: string;
  /** Error message if ok === false */
  error?: string;
}

export interface HopFsWriteResponse extends HopBaseResponse {
  type: 'fs.write';
  ok: boolean;
  error?: string;
}

export interface HopFsDeleteResponse extends HopBaseResponse {
  type: 'fs.delete';
  ok: boolean;
  error?: string;
}

export interface HopFsSearchResponse extends HopBaseResponse {
  type: 'fs.search';
  ok: boolean;
  matches?: string[];
  error?: string;
}

export interface HopWorkspaceOpenResponse extends HopBaseResponse {
  type: 'workspace.open';
  ok: boolean;
  /** Normalized root path, project metadata, etc. */
  workspaceRoot?: string;
  error?: string;
}

export interface HopWorkspaceListResponse extends HopBaseResponse {
  type: 'workspace.list';
  ok: boolean;
  entries?: HopWorkspaceEntry[];
  error?: string;
}

export interface HopWorkspaceEntry {
  path: string;
  kind: 'file' | 'dir' | 'symlink';
  size?: number;
  modified_ms?: number;
}

export interface HopTerminalSpawnResponse extends HopBaseResponse {
  type: 'terminal.spawn';
  ok: boolean;
  /** Internal server-side pid or handle */
  pid?: number;
  error?: string;
}

export interface HopTerminalWriteResponse extends HopBaseResponse {
  type: 'terminal.write';
  ok: boolean;
  error?: string;
}

export interface HopTerminalResizeResponse extends HopBaseResponse {
  type: 'terminal.resize';
  ok: boolean;
  error?: string;
}

export interface HopTerminalKillResponse extends HopBaseResponse {
  type: 'terminal.kill';
  ok: boolean;
  error?: string;
}

export interface HopLspResponse extends HopBaseResponse {
  type: 'lsp.request';
  ok: boolean;
  /** Raw LSP JSON-RPC result or error object */
  result?: unknown;
  error?: string;
}

/** Catch-all protocol-level error */
export interface HopGenericErrorResponse extends HopBaseResponse {
  type: 'error';
  ok: false;
  /** Machine-readable code (optional but recommended) */
  code?: string;
  /** Human-readable description */
  error: string;
}

/* ------------------------------------------------------------------ */
/* Events / notifications (streaming, logs, etc.)                     */
/* ------------------------------------------------------------------ */

export type HopEvent =
  | HopTerminalDataEvent
  | HopTerminalExitEvent
  | HopLspMessageEvent
  | HopLogEvent;

export interface HopTerminalDataEvent {
  type: 'terminal.data';
  id: string; // terminal id
  data: string; // chunk of output
}

export interface HopTerminalExitEvent {
  type: 'terminal.exit';
  id: string;
  code: number | null;
  signal?: string | null;
}

export interface HopLspMessageEvent {
  type: 'lsp.message';
  server: string;
  message: any; // JSON-RPC message
}

export interface HopLogEvent {
  type: 'log';
  level: 'debug' | 'info' | 'warn' | 'error';
  message: string;
  scope?: string;
}

File: apps/hopcoder-shell/src-tauri/src/ipc.rs
use serde::{Deserialize, Serialize};

pub const HOP_IPC_VERSION: u8 = 1;
pub const HOP_EVENT_CHANNEL: &str = "hop://event";

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "kind")]
pub enum HopMessage {
    #[serde(rename = "request")]
    Request(HopRequestMessage),
    #[serde(rename = "response")]
    Response(HopResponseMessage),
    #[serde(rename = "notification")]
    Notification(HopNotificationMessage),
}

#[derive(Serialize, Deserialize, Debug)]
pub struct HopRequestMessage {
    pub v: u8,
    pub id: String,
    pub request: HopRequest,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct HopResponseMessage {
    pub v: u8,
    pub id: String,
    pub response: HopResponse,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct HopNotificationMessage {
    pub v: u8,
    pub event: HopEvent,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type")]
pub enum HopRequest {
    #[serde(rename = "fs.read")]
    FsRead { path: String, root: Option<String> },
    #[serde(rename = "fs.write")]
    FsWrite { path: String, content: String, root: Option<String> },
    #[serde(rename = "fs.delete")]
    FsDelete { path: String, root: Option<String> },
    #[serde(rename = "fs.search")]
    FsSearch { query: String, root: Option<String> },
    #[serde(rename = "workspace.open")]
    WorkspaceOpen { root: String },
    #[serde(rename = "workspace.list")]
    WorkspaceList { root: String },
    #[serde(rename = "terminal.spawn")]
    TerminalSpawn { id: String, shell: Option<String> },
    #[serde(rename = "terminal.write")]
    TerminalWrite { id: String, data: String },
    #[serde(rename = "terminal.resize")]
    TerminalResize { id: String, cols: u32, rows: u32 },
    #[serde(rename = "terminal.kill")]
    TerminalKill { id: String, signal: Option<String> },
    #[serde(rename = "lsp.request")]
    LspRequest { server: String, payload: serde_json::Value },
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type")]
pub enum HopResponse {
    #[serde(rename = "fs.read")]
    FsRead { ok: bool, content: Option<String>, error: Option<String> },
    #[serde(rename = "fs.write")]
    FsWrite { ok: bool, error: Option<String> },
    #[serde(rename = "fs.delete")]
    FsDelete { ok: bool, error: Option<String> },
    #[serde(rename = "fs.search")]
    FsSearch { ok: bool, matches: Option<Vec<String>>, error: Option<String> },
    #[serde(rename = "workspace.open")]
    WorkspaceOpen { ok: bool, #[serde(rename = "workspaceRoot")] workspace_root: Option<String>, error: Option<String> },
    #[serde(rename = "workspace.list")]
    WorkspaceList { ok: bool, entries: Option<Vec<WorkspaceEntry>>, error: Option<String> },
    #[serde(rename = "terminal.spawn")]
    TerminalSpawn { ok: bool, pid: Option<u32>, error: Option<String> },
    #[serde(rename = "terminal.write")]
    TerminalWrite { ok: bool, error: Option<String> },
    #[serde(rename = "terminal.resize")]
    TerminalResize { ok: bool, error: Option<String> },
    #[serde(rename = "terminal.kill")]
    TerminalKill { ok: bool, error: Option<String> },
    #[serde(rename = "lsp.request")]
    LspRequest { ok: bool, result: Option<serde_json::Value>, error: Option<String> },
    #[serde(rename = "error")]
    Error { ok: bool, code: Option<String>, error: String },
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct WorkspaceEntry {
    pub path: String,
    pub kind: String, // "file", "dir", or "symlink"
    pub size: Option<u64>,
    #[serde(rename = "modified_ms")]
    pub modified_ms: Option<i64>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(tag = "type")]
pub enum HopEvent {
    #[serde(rename = "terminal.data")]
    TerminalData { id: String, data: String },
    #[serde(rename = "terminal.exit")]
    TerminalExit { id: String, code: Option<i32>, signal: Option<String> },
    #[serde(rename = "lsp.message")]
    LspMessage { server: String, message: serde_json::Value },
    #[serde(rename = "log")]
    Log { level: String, message: String, scope: Option<String> },
}
